Mathlib/LinearAlgebra/LinearProgramming.lean

```diff
@@ -0,0 +1,83 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.LinearAlgebra.Matrix.ToLin
+import Mathlib.LinearAlgebra.AffineSpace.AffineMap
+
+/-! # Linear programming
+
+Minimizing a linear function on a region defined by linear inequalities.
+
+## Main definitions
+
+ * `LinearProgram` defines a linear program in a form that generalizes "A x ≥ b".
+ * `LinearProgram.feasibles` is the set of all admissible solutions to given linear program.
+ * `LinearProgram.MinAt` defines an optimum solution of given linear program.
+
+-/
+
+/-- Linear program in the form of inequalities with general variables. -/
+structure LinearProgram (K : Type*) {V : Type*} (P : Type*)
+    [LinearOrderedField K] [AddCommGroup V] [Module K V] [AddTorsor V P] where
+  /-- Inequality constraints (given in the form "aᵀx - b ≥ 0") -/
+  constraints : List (P →ᵃ[K] K)
```

Comment:
I am skeptical about making this a `List` rather than a `Set`. One is not actually going to do computation with this type: any meaningful algorithm will require a much more complicated data structure.

Reply:
Happy to hear Martin's take on this, or for someone to click "resolve" as is.

Reply:
If finiteness is really the thing of interest though, you could argue that what we actually want is the type of convex sets built from finitely many half pans; either non-constructively via an existential, or constructively via a Finset and a proof that removing any one element increases the size of the feasible set.

Reply:
Can you induct on Finsets?

Reply:
`Finset.cons_induction`

Reply:
About the second part of your comment, redundant constraints are often a thing that we have to deal with. So we probably shouldn't have the proof obligation that says "removing a constraint strictly increases the feasible set"

Reply:
I remember we talked about something related to the choice `List` vs `Set` in the LP definition, but I don't remember what the discussion was about. Right now, I cannot think of any objection to infinitely many constraints in the LP definition. Of course, many results will be restricted to finite settings.

Reply:
If you allow arbitrarily many conditions, what is the guarantee that you even have a polytope with well defined planes, vertices, faces etc. For example, you could define the set of all tangents to a sphere or an ellipsoid by parametrization as constraints. Would it still be a linear program in any meaningful sense?

Reply:
You are right. Just one comment (not contradicting to your suggestion); all point on the sphere would be vertices.

Reply:
> About the second part of your comment, redundant constraints are often a thing that we have to deal with (because they come from the ambient problem in the LP). So we probably shouldn't have the proof obligation that says "removing a constraint strictly increases the feasible set"

I don't think this claim necessarily holds merit; finsets are not allowed to hold duplicates, and have a proof obligation to this end; but that doesn't stop you from writing `{1, 1, 2, 2}`. You only have to deal with the proof obligation if you call `Finset.mk` directly.

Arguably this "Finset of non-redundant affine constraints" object deserves it's own `Polytope` object.

Reply:
Though another reasonable approach might be to discard the proof obligation and work with the quotient of the type by equality of feasible spaces

Reply:
A redundant LP need not have identical constraints. It just needs to be a linear combination of other constraints (and thereby hold because the other constraints are). I am not sure I see the value in doing, let's say Gaussian elimination on constraints, to extract independent ones makes sense every time an LP is defined. In computer science and also possibly OR, redundant constraints are a fact of life that have to be dealt with algorithmically, and so one often constructs what Mathlib devs might call API lemmas to deal with these cases.

Reply:
When I think about it, having infinitely many conditions is less weird than having infinitely many variables. Given that Mathlib maintainers want the variable space to be general, we should also make the condition space general. Defining conditions as a list (or a multiset, or a finite set) of any type will be bad for LP duality, unless we change the type of `P` along with it.